lib/faraday/request.rb

```diff
@@ -46,14 +46,12 @@ def headers=(hash)
     end
 
     def url(path, params = nil)
-      if path.respond_to? :query
-        if query = path.query
-          path = path.dup
-          path.query = nil
-        end
-      else
-        path, query = path.split('?', 2)
+      path = Utils.URI(path)
```

Comment:
Instead of parsing this with URI, I would simply strip out everything after and including `#` in a string. The fragment identifiers are useless in HTTP requests as they don't actually get sent to the server. So, even if they're supplied, they should be ignored.